### PR TITLE
Add container tags

### PR DESCRIPTION
**Pull request by dineshg13**
**Description:** <Describe what has changed.>

Customers can now set container tags on stats payload computed in DD connector.

sample collector config:
```
  datadog/connector:
    traces:
      resource_attributes_as_container_tags:
        - "k8s.pod.name"
        - "cloud.availability_zone"
```

**Link to tracking Issue:** <Issue number if applicable>

**Testing:** <Describe what testing was performed and which tests were added.>

**Documentation:** <Describe the documentation added.>

### CODE REVIEW

File: connector/datadogconnector/connector.go [hunk 1]
```diff
@@ -80,6 +94,9 @@ func getTraceAgentCfg(cfg TracesConfig, attributesTranslator *attributes.Transla
 	acfg.ComputeStatsBySpanKind = cfg.ComputeStatsBySpanKind
 	acfg.PeerTagsAggregation = cfg.PeerTagsAggregation
 	acfg.PeerTags = cfg.PeerTags
+	if cfg.EnableContainerStats {
+		acfg.Features["enable_cid_stats"] = struct{}{}
```
Comment (jdgumz) [inline connector/datadogconnector/connector.go:98]: :+1:

File: connector/datadogconnector/connector.go [hunk 2]
```diff
@@ -113,6 +130,23 @@ func (c *traceToMetricConnector) Capabilities() consumer.Capabilities {
 
 func (c *traceToMetricConnector) ConsumeTraces(ctx context.Context, traces ptrace.Traces) error {
 	c.agent.Ingest(ctx, traces)
+	if c.enableContainerStats && len(c.containerTagAttrs) > 0 {
```
Comment (jdgumz) [inline connector/datadogconnector/connector.go:133]: Should this happen before the `c.agent.Ingest` call in case of possible race conditions?

File: connector/datadogconnector/connector.go [hunk 3]
```diff
@@ -128,7 +162,19 @@ func (c *traceToMetricConnector) run() {
 			}
 			var mx pmetric.Metrics
 			var err error
+			// Enrich the stats with container tags
+			if c.enableContainerStats && len(c.containerTagCache) > 0 {
+				for _, stat := range stats.Stats {
+					if stat.ContainerID != "" {
+						if tags, ok := c.containerTagCache[stat.ContainerID]; ok {
+							stat.Tags = append(stat.Tags, tags...)
```
Comment (jdgumz) [inline connector/datadogconnector/connector.go:170]: :+1:

File: connector/datadogconnector/connector.go [hunk 4]
```diff
@@ -112,6 +130,27 @@ func (c *traceToMetricConnector) Capabilities() consumer.Capabilities {
...
+			containerID, ok := attrs.Get(semconv.AttributeContainerID)
+			if !ok {
+				continue
+			}
+			ddContainerTags := attributes.ContainerTagsFromResourceAttributes(attrs)
+			for attr := range c.containerTagAttrs {
+				if val, ok := ddContainerTags[attr]; ok {
+					var cacheVal []string
+					if val, ok := c.containerTagCache.Get(containerID.AsString()); ok {
+						cacheVal = val.([]string)
```
Comment (jdgumz) [inline connector/datadogconnector/connector.go:146]: We probably want to do a map for the cache value so that we don't append the same container tag repeatedly?

File: connector/datadogconnector/connector.go [hunk 5]
```diff
@@ -112,6 +128,31 @@ func (c *traceToMetricConnector) Capabilities() consumer.Capabilities {
...
+			if !ok {
+				continue
+			}
+			ddContainerTags := attributes.ContainerTagsFromResourceAttributes(attrs)
+			for attr := range c.resourceAttrs {
+				if val, ok := ddContainerTags[attr]; ok {
+					var cacheVal map[string]struct{}
+					if v, ok := c.containerTagCache.Get(containerID.AsString()); ok {
+						cacheVal = v.(map[string]struct{})
+						cacheVal[fmt.Sprintf("%s:%s", attr, val)] = struct{}{}
```
Comment (jdgumz) [inline connector/datadogconnector/connector.go:145]: nit: I believe `attr + ":" + val` is a little more performant in Go.

File: connector/datadogconnector/connector.go [hunk 6]
```diff
@@ -80,6 +93,9 @@ func getTraceAgentCfg(cfg TracesConfig, attributesTranslator *attributes.Transla
 	acfg.ComputeStatsBySpanKind = cfg.ComputeStatsBySpanKind
 	acfg.PeerTagsAggregation = cfg.PeerTagsAggregation
 	acfg.PeerTags = cfg.PeerTags
+	if len(cfg.ResourceAttributesAsContainerTags) > 0 {
+		acfg.Features["enable_cid_stats"] = struct{}{}
```
Comment (songy23) [inline connector/datadogconnector/connector.go:97]: There is another relevant feature flag `disable_cid_stats` https://github.com/DataDog/datadog-agent/blob/b7ad1ad429262138b4bd198d48874616d98b7417/pkg/trace/stats/concentrator.go#L200

To be safe you can add `delete(acfg.Features, "disable_cid_stats")` too

File: connector/datadogconnector/connector.go (hunk 5 above)
Comment (songy23) [inline connector/datadogconnector/connector.go:145]: nit: can check the existence of the container tag first `_, exists := cacheVal(tag)`

File: connector/datadogconnector/connector.go [hunk 7]
```diff
@@ -128,7 +169,27 @@ func (c *traceToMetricConnector) run() {
...
 			var mx pmetric.Metrics
 			var err error
+			// Enrich the stats with container tags
+			if len(c.resourceAttrs) > 0 {
+				for _, stat := range stats.Stats {
+					if stat.ContainerID != "" {
+						if tags, ok := c.containerTagCache.Get(stat.ContainerID); ok {
+							tagList := tags.(map[string]struct{})
+							// Add unique tags to the stats
+							for _, tag := range stat.Tags {
```
Comment (songy23) [inline connector/datadogconnector/connector.go:179]: Why is this needed? As I understand `stat.Tags` is always empty in concentrator?

File: connector/datadogconnector/connector.go [hunk 8]
```diff
@@ -60,14 +65,22 @@ func newTraceToMetricConnector(set component.TelemetrySettings, cfg component.Co
...
-		in:              in,
-		metricsConsumer: metricsConsumer,
-		exit:            make(chan struct{}),
+		logger:            set.Logger,
+		agent:             datadog.NewAgentWithConfig(ctx, getTraceAgentCfg(cfg.(*Config).Traces, attributesTranslator), in, metricsClient, timingReporter),
+		translator:        trans,
+		in:                in,
+		metricsConsumer:   metricsConsumer,
+		resourceAttrs:     ddtags,
+		containerTagCache: cache.New(cache.DefaultExpiration, cache.DefaultExpiration),
```
Comment (songy23) [inline connector/datadogconnector/connector.go:82]: What does setting `DefaultExpiration` do, e.g. never expire? Maybe add a comment here

**Review by songy23 — COMMENTED** [review]
Needs `make gotidy`

File: connector/datadogconnector/connector.go (hunk 7 above)
Comment (dineshg13) [reply connector/datadogconnector/connector.go:179]: if it empty - the loop would be noop.

File: connector/datadogconnector/config.go [hunk 9]
```diff
@@ -63,6 +63,9 @@ type TracesConfig struct {
 	// TraceBuffer specifies the number of Datadog Agent TracerPayloads to buffer before dropping.
 	// The default value is 1000.
 	TraceBuffer int `mapstructure:"trace_buffer"`
+
+	// ResourceAttributesAsContainerTags specifies the list of resource attributes to be used as container tags.
+	ResourceAttributesAsContainerTags []string `mapstructure:"resource_attributes_as_container_tags"`
```
Comment (jdgumz) [inline connector/datadogconnector/config.go:68]: Can you share an example configuration, perhaps in the PR description?

File: connector/datadogconnector/connector.go [hunk 10]
```diff
@@ -46,6 +52,13 @@ type traceToMetricConnector struct {
 
 var _ component.Component = (*traceToMetricConnector)(nil) // testing that the connectorImp properly implements the type Component interface
 
+// cacheExpiration is the time after which a container tag cache entry will expire
+// and be removed from the cache.
+var cacheExpiration = time.Minute * 5
```
Comment (jdgumz) [inline connector/datadogconnector/connector.go:57]: Do we want to make this more configurable?

Comment (dineshg13) [reply connector/datadogconnector/connector.go:57]: This is temporary. We should get rid of it next quarter, where concentrator picks up tags from the incoming payload.

File: connector/datadogconnector/config.go (hunk 9 above)
Comment (songy23) [reply connector/datadogconnector/config.go:68]: https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/connector/datadogconnector/README.md#configurations and https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/connector/datadogconnector/examples/config.yaml#L25 need to be updated.

File: connector/datadogconnector/connector.go [hunk 11]
```diff
@@ -112,6 +137,31 @@ func (c *traceToMetricConnector) Capabilities() consumer.Capabilities {
 }
 
 func (c *traceToMetricConnector) ConsumeTraces(ctx context.Context, traces ptrace.Traces) error {
+	if len(c.resourceAttrs) > 0 {
```
Comment (jdgumz) [inline connector/datadogconnector/connector.go:140]: Should we put some of this into a separate function or two for readability?

File: connector/datadogconnector/connector.go [hunk 12]
```diff
@@ -128,7 +178,27 @@ func (c *traceToMetricConnector) run() {
 			}
 			var mx pmetric.Metrics
 			var err error
+			// Enrich the stats with container tags
```
Comment (jdgumz) [inline connector/datadogconnector/connector.go:181]: Same comment here as above: maybe make a function for this?

**Review by jdgumz — APPROVED** [review]

File: connector/datadogconnector/connector.go [hunk 13]
```diff
@@ -60,14 +73,22 @@ func newTraceToMetricConnector(set component.TelemetrySettings, cfg component.Co
 		return nil, fmt.Errorf("failed to create metrics translator: %w", err)
 	}
 
+	ctags := make(map[string]string, len(cfg.(*Config).Traces.ResourceAttributesAsContainerTags))
```
[the remark this answers is not in the document]
Comment (dineshg13) [reply connector/datadogconnector/connector.go:76]: `ContainerTagFromAttributes` works on map[string]string , hence we had to use same

**Review by songy23 — APPROVED** [review]
LGTM
